sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -497,23 +496,33 @@ public void setWindowedWrites(boolean windowedWrites) {
      * idempotent, as it may be executed multiple times in the case of failure or for redundancy. It
      * is a best practice to attempt to try to make this method atomic.
      *
+     * <p>Returns the set of temporary files generated. Callers must call
+     * {@link #removeTemporaryFiles(Set)} to cleanup these files.
+     *
      * @param writerResults the results of writes (FileResult).
      */
-    public void finalize(Iterable<FileResult<DestinationT>> writerResults) throws Exception {
-      // Collect names of temporary files and rename them.
+     public Set<ResourceId> finalize(
+         Iterable<FileResult<DestinationT>> writerResults) throws Exception {
+      // Collect names of temporary files and copies them.
       Map<ResourceId, ResourceId> outputFilenames = buildOutputFilenames(writerResults);
       copyToOutputFiles(outputFilenames);
+       return outputFilenames.keySet();
+    }
 
-      // Optionally remove temporary files.
-      // We remove the entire temporary directory, rather than specifically removing the files
-      // from writerResults, because writerResults includes only successfully completed bundles,
-      // and we'd like to clean up the failed ones too.
-      // Note that due to GCS eventual consistency, matching files in the temp directory is also
-      // currently non-perfect and may fail to delete some files.
-      //
-      // When windows or triggers are specified, files are generated incrementally so deleting
-      // the entire directory in finalize is incorrect.
-      removeTemporaryFiles(outputFilenames.keySet(), !windowedWrites);
+    /*
+     * Remove temporary files after finalization.
+     *
+     * <p>We remove the entire temporary directory, rather than specifically removing the files from
```

Comment:
Here's how I understand the restrictions on removing the temp directory:

* We should definitely not remove the temp directory in finalize(), because we might be finalizing multiple destinations writing to the same directory - the directory can be removed only after all of them have been finalized, otherwise finalizing one destination would remove temp files produced by another destination.
* We should also not remove the temp directory _at all_ when writing an unbounded collection, because new data may appear in the collection and we'll need the directory again.

I think currently the comment is misleading on a few counts:
* it says we remove the entire directory, while really we do it only sometimes;
* it says that the reason why removal is in removeTemporaryFiles() rather than finalize() has to do with triggers, but actually it has to do with multiple destinations

Also I believe the condition should be not "!windowedWrites", but "is the input collection unbounded". When doing windowed writes of a bounded collection, I believe we should still remove the temp directory.